cpp/inc/bond/protocol/compact_binary.h

```diff
@@ -381,129 +381,231 @@ class CompactBinaryReader
+    void SkipStructV1()
+    {
+        BOOST_ASSERT(v1 == _version);
 
-            case bond::BT_BOOL:
-            case bond::BT_UINT8:
-            case bond::BT_INT8:
-                _input.Skip(sizeof(uint8_t));
-                break;
+        for (;;)
+        {
+            ReadStructBegin();
 
-            case bond::BT_UINT64:
-            case bond::BT_UINT32:
-            case bond::BT_UINT16:
-            case bond::BT_INT64:
-            case bond::BT_INT32:
-            case bond::BT_INT16:
+            uint16_t     id;
+            BondDataType field_type;
+
+            for (ReadFieldBegin(field_type, id);
+                    field_type != bond::BT_STOP && field_type != bond::BT_STOP_BASE;
+                    ReadFieldEnd(), ReadFieldBegin(field_type, id))
             {
-                uint64_t value;
-                Read(value);
-                break;
+                SkipType(field_type);
             }
-            default:
-                SkipComplex(type);
+
+            ReadStructEnd();
+
+            if (field_type == bond::BT_STOP)
                 break;
         }
     }
 
-protected:
-    void SkipComplex(BondDataType type)
+    void SkipStructV2()
     {
-        switch (type)
+        BOOST_ASSERT(v2 == _version);
+
+        uint32_t length;
+        Read(length);
+        _input.Skip(length);
+    }
+
+    template <BondDataType T>
+    typename boost::enable_if_c<(T == BT_STRUCT)>::type
+    SkipType()
+    {
+        if (v2 == _version)
         {
-            case bond::BT_STRING:
-            {
-                uint32_t length;
+            SkipStructV2();
+        }
+        else
+        {
+            SkipStructV1();
+        }
+    }
 
-                Read(length);
-                _input.Skip(length);
-                break;
+    template <BondDataType T>
+    typename boost::enable_if_c<(T == BT_STRUCT)>::type
+    SkipType(uint32_t size)
+    {
+        if (v2 == _version)
+        {
+            for (uint32_t i = 0; i < size; ++i)
+            {
+                SkipStructV2();
             }
-            case bond::BT_WSTRING:
+        }
+        else
+        {
+            for (uint32_t i = 0; i < size; ++i)
             {
-                uint32_t length;
+                SkipStructV1();
+            }
+        }
+    }
 
-                Read(length);
-                _input.Skip(length * sizeof(uint16_t));
+    template <BondDataType T>
+    typename boost::enable_if_c<(T == BT_UINT16 || T == BT_UINT32 || T == BT_UINT64
+                                || T == BT_INT16 || T == BT_INT32 || T == BT_INT64
+                                || T == BT_STRING || T == BT_WSTRING
+                                || T == BT_SET || T == BT_LIST || T == BT_MAP)>::type
+    SkipType(uint32_t size)
+    {
+        for (uint32_t i = 0; i < size; ++i)
+        {
+            SkipType<T>();
+        }
+    }
+
+    template <typename... Args>
+    void SkipType(BondDataType type, Args&&... args)
```

Comment:
Does this ever get called with non-empty `args`?

Reply:
Yeah, this is just a trick to share the same code for `SkipType(type)` and `SkipType(type, size)`.

Reply:
👌